File: server/createQuoteImage.py
#!/usr/bin/env python3
"""
EasySignals Quote of the Day – Premium Design v3
Dark background, gold/white text, candlestick chart, glow curve
Cinzel + Playfair Display fonts
"""
import sys
import os
import math
import hashlib
from datetime import date
from PIL import Image, ImageDraw, ImageFont

SIZE = 1080

GOLD = (212, 175, 55)
GOLD_LIGHT = (255, 215, 80)
WHITE = (255, 255, 255)
BG_DARK = (5, 18, 12)
GREEN_GLOW = (0, 255, 100)

FONT_DIR = "/home/ubuntu/webdev-static-assets/fonts"
LOGO_PATH = "/home/ubuntu/webdev-static-assets/easysignals_logo_white.png"
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

# Background variants (existing)
BG_VARIANTS = [
    os.path.join(SCRIPT_DIR, "quote_bg.png"),
    os.path.join(SCRIPT_DIR, "quote_bg_blue.png"),
    os.path.join(SCRIPT_DIR, "quote_bg_darkgreen.png"),
]

def get_daily_bg():
    today_str = date.today().isoformat()
    idx = int(hashlib.md5(today_str.encode()).hexdigest(), 16) % len(BG_VARIANTS)
    return BG_VARIANTS[idx]

def load_font(name, size):
    path = os.path.join(FONT_DIR, name)
    if os.path.exists(path):
        return ImageFont.truetype(path, size)
    # Fallback to system fonts
    for fallback in ["/usr/share/fonts/truetype/liberation/LiberationSerif-Bold.ttf",
                     "/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf"]:
        if os.path.exists(fallback):
            return ImageFont.truetype(fallback, size)
    return ImageFont.load_default()

def draw_glow_curve(img):
    """Glowing green curve on the right side"""
    overlay = Image.new('RGBA', (SIZE, SIZE), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)
    points = []
    for i in range(300):
        t = i / 299
        x = int(SIZE * 0.28 + SIZE * 0.72 * t)
        y = int(SIZE * 0.98 - SIZE * 0.88 * (t ** 1.3))
        points.append((x, y))
    for width, alpha, color in [
        (22, 20, (0, 255, 100)),
        (14, 45, (0, 255, 120)),
        (6, 130, (80, 255, 150)),
        (2, 220, (200, 255, 220)),
    ]:
        for i in range(len(points) - 1):
            draw.line([points[i], points[i+1]], fill=(*color, alpha), width=width)
    img.paste(overlay, (0, 0), overlay)

def draw_candlesticks(img):
    overlay = Image.new('RGBA', (SIZE, SIZE), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)
    # Left side (smaller)
    for i, (x, top, bt, bb, green) in enumerate([
        (55, 830, 790, 750, True), (100, 790, 750, 710, False),
        (145, 760, 715, 670, True), (190, 720, 670, 620, True),
        (235, 680, 625, 575, False), (280, 640, 580, 530, True),
    ]):
        c = (0, 180, 80, 70) if green else (200, 60, 60, 70)
        draw.line([(x, top), (x, max(bt,bb) + 40)], fill=c, width=2)
        draw.rectangle([x-7, min(bt,bb), x+7, max(bt,bb)], fill=c)
    # Right side (larger)
    for x, top, bt, bb, green in [
        (820, 620, 420, 510, True), (870, 490, 300, 390, True),
        (920, 370, 190, 280, False), (965, 280, 110, 200, True),
        (1010, 200, 50, 130, True),
    ]:
        c = (0, 220, 100, 110) if green else (220, 80, 80, 110)
        draw.line([(x, top), (x, max(bt,bb) + 40)], fill=c, width=3)
        draw.rectangle([x-13, min(bt,bb), x+13, max(bt,bb)], fill=c)
    img.paste(overlay, (0, 0), overlay)

def draw_logo(img):
    logo_path = LOGO_PATH
    if not os.path.exists(logo_path):
        # Try script dir
        logo_path = os.path.join(SCRIPT_DIR, "easysignals_logo.png")
    if not os.path.exists(logo_path):
        return
    logo = Image.open(logo_path).convert("RGBA")
    logo_w = 300
    logo_h = int(logo.height * logo_w / logo.width)
    logo = logo.resize((logo_w, logo_h), Image.LANCZOS)
    x = (SIZE - logo_w) // 2
    img.paste(logo, (x, 48), logo)

def wrap_text_lines(text, font, max_width, draw):
    words = text.split()
    lines = []
    cur = []
    for word in words:
        test = ' '.join(cur + [word])
        bbox = draw.textbbox((0, 0), test, font=font)
        if bbox[2] - bbox[0] <= max_width and cur:
            cur.append(word)
        elif not cur:
            cur.append(word)
        else:
            lines.append(' '.join(cur))
            cur = [word]
    if cur:
        lines.append(' '.join(cur))
    return lines

def create_quote_image(quote: str, author: str, output_path: str):
    # Background
    bg_path = get_daily_bg()
    if os.path.exists(bg_path):
        img = Image.open(bg_path).convert("RGBA")
        img = img.resize((SIZE, SIZE), Image.LANCZOS)
        # Darken slightly for text readability
        dark = Image.new("RGBA", (SIZE, SIZE), (0, 0, 0, 60))
        img.alpha_composite(dark)
    else:
        img = Image.new('RGBA', (SIZE, SIZE), (*BG_DARK, 255))
        # Simple gradient
        draw_tmp = ImageDraw.Draw(img)
        for y in range(SIZE):
            t = y / SIZE
            r = int(5 + 8 * t); g = int(18 + 20 * t); b = int(12 + 10 * t)
            draw_tmp.line([(0, y), (SIZE, y)], fill=(r, g, b, 255))

    # Candlesticks + glow
    draw_candlesticks(img)
    draw_glow_curve(img)

    # Logo
    draw_logo(img)

    # Fonts
    f_cinzel_badge  = load_font("Cinzel.ttf", 30)
    f_cinzel_author = load_font("Cinzel.ttf", 34)
    f_play_lg       = load_font("PlayfairDisplay.ttf", 90)
    f_play_md       = load_font("PlayfairDisplay.ttf", 74)
    f_play_sm       = load_font("PlayfairDisplay.ttf", 58)
    f_play_italic   = load_font("PlayfairDisplay-Italic.ttf", 78)
    f_play_italic_sm= load_font("PlayfairDisplay-Italic.ttf", 62)
    f_play_qmark    = load_font("PlayfairDisplay.ttf", 130)

    draw = ImageDraw.Draw(img)

    # ── TITLE BADGE ───────────────────────────────────────────────────────────
    badge_text = "QUOTE OF THE DAY"
    bb = draw.textbbox((0, 0), badge_text, font=f_cinzel_badge)
    bw, bh = bb[2]-bb[0], bb[3]-bb[1]
    px, py = 38, 14
    badge_x = (SIZE - bw) // 2 - px
    badge_y = 152
    badge_rect = [badge_x, badge_y, badge_x + bw + px*2, badge_y + bh + py*2]

    # Badge background + border
    badge_overlay = Image.new('RGBA', (SIZE, SIZE), (0, 0, 0, 0))
    bd = ImageDraw.Draw(badge_overlay)
    bd.rounded_rectangle(badge_rect, radius=6, fill=(5, 15, 10, 210), outline=(*GOLD, 255), width=2)
    # Corner diamonds
    for cx, cy in [(badge_x, badge_y), (badge_x + bw + px*2, badge_y),
                   (badge_x, badge_y + bh + py*2), (badge_x + bw + px*2, badge_y + bh + py*2)]:
        bd.polygon([(cx, cy-6), (cx+6, cy), (cx, cy+6), (cx-6, cy)], fill=(*GOLD, 200))
    # Side lines
    mid_y = badge_y + (bh + py*2) // 2
    bd.line([(badge_x - 70, mid_y), (badge_x - 8, mid_y)], fill=(*GOLD, 140), width=1)
    bd.line([(badge_x + bw + px*2 + 8, mid_y), (badge_x + bw + px*2 + 70, mid_y)], fill=(*GOLD, 140), width=1)
    img.alpha_composite(badge_overlay)
    draw = ImageDraw.Draw(img)
    draw.text(((SIZE - bw) // 2, badge_y + py), badge_text, font=f_cinzel_badge, fill=GOLD)

    # ── OPENING QUOTE MARK ────────────────────────────────────────────────────
    qm_y = 230
    qm_bb = draw.textbbox((0, 0), "\u201c", font=f_play_qmark)
    qm_w = qm_bb[2] - qm_bb[0]
    draw.text(((SIZE - qm_w) // 2, qm_y), "\u201c", font=f_play_qmark, fill=GOLD)

    # Decorative lines beside quote mark
    line_y = qm_y + 55
    draw.line([(70, line_y), (SIZE//2 - 90, line_y)], fill=(*GOLD, 100), width=1)
    draw.line([(SIZE//2 + 90, line_y), (SIZE - 70, line_y)], fill=(*GOLD, 100), width=1)

    # ── QUOTE TEXT ────────────────────────────────────────────────────────────
    margin = 72
    max_w = SIZE - 2 * margin
    words = quote.split()

    # Split: last ~30% of words get italic gold treatment
    italic_count = max(2, len(words) // 3)
    main_words = words[:-italic_count] if len(words) > italic_count + 2 else words
    italic_words = words[-italic_count:] if len(words) > italic_count + 2 else []

    main_text = ' '.join(main_words).upper()
    italic_text = ' '.join(italic_words)

    # Choose font size based on length
    total_chars = len(quote)
    if total_chars <= 60:
        f_main, f_ital, lh = f_play_lg, f_play_italic, 105
    elif total_chars <= 100:
        f_main, f_ital, lh = f_play_md, f_play_italic, 88
    else:
        f_main, f_ital, lh = f_play_sm, f_play_italic_sm, 72

    main_lines = wrap_text_lines(main_text, f_main, max_w, draw)
    italic_lines = wrap_text_lines(italic_text, f_ital, max_w, draw) if italic_text else []

    total_lines = len(main_lines) + len(italic_lines)
    text_start_y = 310
    available = 840 - text_start_y
    total_h = total_lines * lh
    y = text_start_y + max(0, (available - total_h) // 2)

    # Draw main lines – alternating white / gold
    for i, line in enumerate(main_lines):
        color = GOLD if i % 2 == 1 else WHITE
        bb2 = draw.textbbox((0, 0), line, font=f_main)
        w = bb2[2] - bb2[0]
        x = (SIZE - w) // 2
        # Shadow
        draw.text((x+3, y+3), line, font=f_main, fill=(0, 0, 0))
        draw.text((x, y), line, font=f_main, fill=color)
        y += lh

    # Draw italic lines in gold
    for line in italic_lines:
        bb3 = draw.textbbox((0, 0), line, font=f_ital)
        w = bb3[2] - bb3[0]
        x = (SIZE - w) // 2
        draw.text((x+3, y+3), line, font=f_ital, fill=(0, 0, 0))
        draw.text((x, y), line, font=f_ital, fill=GOLD_LIGHT)
        y += lh

    # ── CLOSING QUOTE MARK ────────────────────────────────────────────────────
    cqm_bb = draw.textbbox((0, 0), "\u201d", font=f_play_qmark)
    cqm_w = cqm_bb[2] - cqm_bb[0]
    draw.text(((SIZE - cqm_w) // 2, y + 5), "\u201d", font=f_play_qmark, fill=GOLD)

    # ── AUTHOR ────────────────────────────────────────────────────────────────
    author_y = y + 75
    cx = SIZE // 2
    draw.line([(cx - 130, author_y), (cx + 130, author_y)], fill=(*GOLD, 160), width=1)

    author_str = f"\u2013 {author.upper()}"
    ab = draw.textbbox((0, 0), author_str, font=f_cinzel_author)
    aw = ab[2] - ab[0]
    draw.text(((SIZE - aw) // 2 + 2, author_y + 18 + 2), author_str, font=f_cinzel_author, fill=(0, 0, 0))
    draw.text(((SIZE - aw) // 2, author_y + 18), author_str, font=f_cinzel_author, fill=WHITE)

    # Save
    final = img.convert("RGB")
    final.save(output_path, "PNG", quality=95)
    print(f"✅ Quote image saved: {output_path}")

if __name__ == "__main__":
    q = sys.argv[1] if len(sys.argv) > 1 else "The most important thing in making money is not letting your losses get out of hand."
    a = sys.argv[2] if len(sys.argv) > 2 else "Marty Schwartz"
    o = sys.argv[3] if len(sys.argv) > 3 else "/tmp/quote_test_v3.png"
    create_quote_image(q, a, o)

File: server/createQuoteImage.ts
/**
 * EasySignals Quote of the Day – Node.js Image Generator
 * Replaces the Python/Pillow version for production compatibility.
 * Uses @napi-rs/canvas for 2D drawing and sharp for image processing.
 */
import { createCanvas, loadImage, GlobalFonts } from "@napi-rs/canvas";
import sharp from "sharp";
import { join, dirname } from "path";
import { fileURLToPath } from "url";
import { existsSync, writeFileSync, unlinkSync } from "fs";
import { tmpdir } from "os";
import { createHash } from "crypto";

const SIZE = 1080;

// Colors
const GOLD: [number, number, number] = [212, 175, 55];
const GOLD_LIGHT: [number, number, number] = [255, 215, 80];
const WHITE: [number, number, number] = [255, 255, 255];
const BG_DARK: [number, number, number] = [5, 18, 12];

// Use paths relative to this file so they work in both dev and production
const _serverDir = dirname(fileURLToPath(import.meta.url));
const FONT_DIR = join(_serverDir, "fonts");
const LOGO_PATH = join(_serverDir, "assets", "logo_white.png");

function rgba(r: number, g: number, b: number, a = 1): string {
  return `rgba(${r},${g},${b},${a})`;
}
function rgb(c: [number, number, number]): string {
  return `rgb(${c[0]},${c[1]},${c[2]})`;
}

function registerFonts() {
  const fonts = [
    { file: "Cinzel.ttf", family: "Cinzel" },
    { file: "PlayfairDisplay.ttf", family: "PlayfairDisplay" },
    { file: "PlayfairDisplay-Italic.ttf", family: "PlayfairDisplayItalic" },
  ];
  for (const { file, family } of fonts) {
    const path = join(FONT_DIR, file);
    if (existsSync(path)) {
      GlobalFonts.registerFromPath(path, family);
    }
  }
  // Fallback system fonts
  const fallbacks = [
    "/usr/share/fonts/truetype/liberation/LiberationSerif-Bold.ttf",
    "/usr/share/fonts/truetype/liberation/LiberationSans-Bold.ttf",
  ];
  for (const f of fallbacks) {
    if (existsSync(f)) {
      GlobalFonts.registerFromPath(f, "FallbackSerif");
    }
  }
}

function fontSpec(family: string, size: number, italic = false): string {
  const hasCinzel = existsSync(join(FONT_DIR, "Cinzel.ttf"));
  const hasPlayfair = existsSync(join(FONT_DIR, "PlayfairDisplay.ttf"));

  if (family === "Cinzel") {
    return `bold ${size}px ${hasCinzel ? "Cinzel" : "FallbackSerif, serif"}`;
  }
  if (family === "PlayfairDisplay") {
    if (italic) {
      return `italic ${size}px ${hasPlayfair ? "PlayfairDisplayItalic, PlayfairDisplay" : "FallbackSerif, serif"}`;
    }
    return `bold ${size}px ${hasPlayfair ? "PlayfairDisplay" : "FallbackSerif, serif"}`;
  }
  return `bold ${size}px serif`;
}

function wrapTextLines(
  ctx: ReturnType<ReturnType<typeof createCanvas>["getContext"]>,
  text: string,
  font: string,
  maxWidth: number
): string[] {
  ctx.font = font;
  const words = text.split(" ");
  const lines: string[] = [];
  let cur: string[] = [];

  for (const word of words) {
    const test = [...cur, word].join(" ");
    const w = ctx.measureText(test).width;
    if (w <= maxWidth && cur.length > 0) {
      cur.push(word);
    } else if (cur.length === 0) {
      cur.push(word);
    } else {
      lines.push(cur.join(" "));
      cur = [word];
    }
  }
  if (cur.length > 0) lines.push(cur.join(" "));
  return lines;
}

function drawGlowCurve(ctx: ReturnType<ReturnType<typeof createCanvas>["getContext"]>) {
  const points: [number, number][] = [];
  for (let i = 0; i < 300; i++) {
    const t = i / 299;
    const x = SIZE * 0.28 + SIZE * 0.72 * t;
    const y = SIZE * 0.98 - SIZE * 0.88 * Math.pow(t, 1.3);
    points.push([x, y]);
  }

  const layers: [number, number, [number, number, number]][] = [
    [22, 0.08, [0, 255, 100]],
    [14, 0.18, [0, 255, 120]],
    [6, 0.51, [80, 255, 150]],
    [2, 0.86, [200, 255, 220]],
  ];

  for (const [width, alpha, color] of layers) {
    ctx.save();
    ctx.strokeStyle = rgba(color[0], color[1], color[2], alpha);
    ctx.lineWidth = width;
    ctx.beginPath();
    ctx.moveTo(points[0][0], points[0][1]);
    for (let i = 1; i < points.length; i++) {
      ctx.lineTo(points[i][0], points[i][1]);
    }
    ctx.stroke();
    ctx.restore();
  }
}

function drawCandlesticks(ctx: ReturnType<ReturnType<typeof createCanvas>["getContext"]>) {
  // Left side (smaller)
  const leftCandles: [number, number, number, number, boolean][] = [
    [55, 830, 790, 750, true], [100, 790, 750, 710, false],
    [145, 760, 715, 670, true], [190, 720, 670, 620, true],
    [235, 680, 625, 575, false], [280, 640, 580, 530, true],
  ];
  for (const [x, top, bt, bb, green] of leftCandles) {
    const c = green ? rgba(0, 180, 80, 0.27) : rgba(200, 60, 60, 0.27);
    ctx.save();
    ctx.strokeStyle = c;
    ctx.lineWidth = 2;
    ctx.beginPath();
    ctx.moveTo(x, top);
    ctx.lineTo(x, Math.max(bt, bb) + 40);
    ctx.stroke();
    ctx.fillStyle = c;
    ctx.fillRect(x - 7, Math.min(bt, bb), 14, Math.abs(bt - bb));
    ctx.restore();
  }

  // Right side (larger)
  const rightCandles: [number, number, number, number, boolean][] = [
    [820, 620, 420, 510, true], [870, 490, 300, 390, true],
    [920, 370, 190, 280, false], [965, 280, 110, 200, true],
    [1010, 200, 50, 130, true],
  ];
  for (const [x, top, bt, bb, green] of rightCandles) {
    const c = green ? rgba(0, 220, 100, 0.43) : rgba(220, 80, 80, 0.43);
    ctx.save();
    ctx.strokeStyle = c;
    ctx.lineWidth = 3;
    ctx.beginPath();
    ctx.moveTo(x, top);
    ctx.lineTo(x, Math.max(bt, bb) + 40);
    ctx.stroke();
    ctx.fillStyle = c;
    ctx.fillRect(x - 13, Math.min(bt, bb), 26, Math.abs(bt - bb));
    ctx.restore();
  }
}

async function loadBackground(backgroundUrl?: string): Promise<Buffer> {
  // 1. DALL-E 3 URL
  if (backgroundUrl) {
    try {
      console.log(`[createQuoteImage] Lade Hintergrund: ${backgroundUrl.slice(0, 60)}...`);
      const resp = await fetch(backgroundUrl, { signal: AbortSignal.timeout(30000) });
      if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
      const arrayBuf = await resp.arrayBuffer();
      const buf = await sharp(Buffer.from(arrayBuf))
        .resize(SIZE, SIZE, { fit: "cover" })
        .toBuffer();
      console.log("[createQuoteImage] Hintergrund geladen ✅");
      return buf;
    } catch (e) {
      console.warn(`[createQuoteImage] Hintergrund fehlgeschlagen: ${e} – Fallback`);
    }
  }

  // 2. Static fallback backgrounds
  const scriptDir = join(process.cwd(), "server");
  const bgVariants = [
    join(scriptDir, "quote_bg.png"),
    join(scriptDir, "quote_bg_blue.png"),
    join(scriptDir, "quote_bg_darkgreen.png"),
  ];
  const today = new Date().toISOString().slice(0, 10);
  const idx = parseInt(createHash("md5").update(today).digest("hex"), 16) % bgVariants.length;
  const bgPath = bgVariants[idx];
  if (existsSync(bgPath)) {
    return await sharp(bgPath).resize(SIZE, SIZE, { fit: "cover" }).toBuffer();
  }

  // 3. Solid dark gradient fallback
  const gradCanvas = createCanvas(SIZE, SIZE);
  const gCtx = gradCanvas.getContext("2d");
  const grad = gCtx.createLinearGradient(0, 0, 0, SIZE);
  grad.addColorStop(0, rgb(BG_DARK));
  grad.addColorStop(1, "rgb(13,38,22)");
  gCtx.fillStyle = grad;
  gCtx.fillRect(0, 0, SIZE, SIZE);
  return gradCanvas.toBuffer("image/png");
}

export async function createQuoteImage(
  quote: string,
  author: string,
  outputPath: string,
  backgroundUrl?: string
): Promise<void> {
  registerFonts();

  const canvas = createCanvas(SIZE, SIZE);
  const ctx = canvas.getContext("2d");

  // ── BACKGROUND ──────────────────────────────────────────────────────────────
  const bgBuf = await loadBackground(backgroundUrl);
  const bgImg = await loadImage(bgBuf);
  ctx.drawImage(bgImg, 0, 0, SIZE, SIZE);

  // Dark overlay for readability
  ctx.fillStyle = rgba(0, 0, 0, 0.39);
  ctx.fillRect(0, 0, SIZE, SIZE);

  // ── DECORATIVE ELEMENTS ─────────────────────────────────────────────────────
  drawCandlesticks(ctx);
  drawGlowCurve(ctx);

  // ── LOGO ────────────────────────────────────────────────────────────────────
  if (existsSync(LOGO_PATH)) {
    try {
      const logo = await loadImage(LOGO_PATH);
      const logoW = 300;
      const logoH = Math.round(logo.height * logoW / logo.width);
      const logoX = (SIZE - logoW) / 2;
      ctx.drawImage(logo, logoX, 48, logoW, logoH);
    } catch (e) {
      console.warn("[createQuoteImage] Logo konnte nicht geladen werden:", e);
    }
  }

  // ── BADGE ───────────────────────────────────────────────────────────────────
  const badgeText = "QUOTE OF THE DAY";
  const badgeFont = fontSpec("Cinzel", 30);
  ctx.font = badgeFont;
  const bm = ctx.measureText(badgeText);
  const bw = bm.width;
  const bh = 36;
  const px = 38, py = 14;
  const badgeX = (SIZE - bw) / 2 - px;
  const badgeY = 152;
  const badgeW = bw + px * 2;
  const badgeH = bh + py * 2;

  // Badge background
  ctx.save();
  ctx.fillStyle = rgba(5, 15, 10, 0.82);
  ctx.strokeStyle = rgb(GOLD);
  ctx.lineWidth = 2;
  const r = 6;
  ctx.beginPath();
  ctx.moveTo(badgeX + r, badgeY);
  ctx.lineTo(badgeX + badgeW - r, badgeY);
  ctx.arcTo(badgeX + badgeW, badgeY, badgeX + badgeW, badgeY + r, r);
  ctx.lineTo(badgeX + badgeW, badgeY + badgeH - r);
  ctx.arcTo(badgeX + badgeW, badgeY + badgeH, badgeX + badgeW - r, badgeY + badgeH, r);
  ctx.lineTo(badgeX + r, badgeY + badgeH);
  ctx.arcTo(badgeX, badgeY + badgeH, badgeX, badgeY + badgeH - r, r);
  ctx.lineTo(badgeX, badgeY + r);
  ctx.arcTo(badgeX, badgeY, badgeX + r, badgeY, r);
  ctx.closePath();
  ctx.fill();
  ctx.stroke();
  ctx.restore();

  // Corner diamonds
  const corners: [number, number][] = [
    [badgeX, badgeY], [badgeX + badgeW, badgeY],
    [badgeX, badgeY + badgeH], [badgeX + badgeW, badgeY + badgeH],
  ];
  for (const [cx, cy] of corners) {
    ctx.save();
    ctx.fillStyle = rgba(GOLD[0], GOLD[1], GOLD[2], 0.78);
    ctx.beginPath();
    ctx.moveTo(cx, cy - 6);
    ctx.lineTo(cx + 6, cy);
    ctx.lineTo(cx, cy + 6);
    ctx.lineTo(cx - 6, cy);
    ctx.closePath();
    ctx.fill();
    ctx.restore();
  }

  // Side lines
  const midY = badgeY + badgeH / 2;
  ctx.save();
  ctx.strokeStyle = rgba(GOLD[0], GOLD[1], GOLD[2], 0.55);
  ctx.lineWidth = 1;
  ctx.beginPath();
  ctx.moveTo(badgeX - 70, midY);
  ctx.lineTo(badgeX - 8, midY);
  ctx.moveTo(badgeX + badgeW + 8, midY);
  ctx.lineTo(badgeX + badgeW + 70, midY);
  ctx.stroke();
  ctx.restore();

  // Badge text
  ctx.font = badgeFont;
  ctx.fillStyle = rgb(GOLD);
  ctx.fillText(badgeText, (SIZE - bw) / 2, badgeY + py + bh * 0.78);

  // ── OPENING QUOTE MARK ──────────────────────────────────────────────────────
  const qmFont = fontSpec("PlayfairDisplay", 130);
  ctx.font = qmFont;
  const qmW = ctx.measureText("\u201c").width;
  const qmY = 230;
  ctx.fillStyle = rgb(GOLD);
  ctx.fillText("\u201c", (SIZE - qmW) / 2, qmY + 100);

  // Decorative lines beside quote mark
  const lineY = qmY + 55;
  ctx.save();
  ctx.strokeStyle = rgba(GOLD[0], GOLD[1], GOLD[2], 0.39);
  ctx.lineWidth = 1;
  ctx.beginPath();
  ctx.moveTo(70, lineY);
  ctx.lineTo(SIZE / 2 - 90, lineY);
  ctx.moveTo(SIZE / 2 + 90, lineY);
  ctx.lineTo(SIZE - 70, lineY);
  ctx.stroke();
  ctx.restore();

  // ── QUOTE TEXT ──────────────────────────────────────────────────────────────
  const margin = 72;
  const maxW = SIZE - 2 * margin;
  const words = quote.split(" ");
  const totalChars = quote.length;

  let mainFontSize: number, italicFontSize: number, lineH: number;
  if (totalChars <= 60) {
    mainFontSize = 90; italicFontSize = 78; lineH = 105;
  } else if (totalChars <= 100) {
    mainFontSize = 74; italicFontSize = 78; lineH = 88;
  } else {
    mainFontSize = 58; italicFontSize = 62; lineH = 72;
  }

  const mainFont = fontSpec("PlayfairDisplay", mainFontSize);
  const italicFont = fontSpec("PlayfairDisplay", italicFontSize, true);

  const italicCount = Math.max(2, Math.floor(words.length / 3));
  const mainWords = words.length > italicCount + 2 ? words.slice(0, -italicCount) : words;
  const italicWords = words.length > italicCount + 2 ? words.slice(-italicCount) : [];

  const mainText = mainWords.join(" ").toUpperCase();
  const italicText = italicWords.join(" ");

  const mainLines = wrapTextLines(ctx, mainText, mainFont, maxW);
  const italicLines = italicText ? wrapTextLines(ctx, italicText, italicFont, maxW) : [];

  const totalLines = mainLines.length + italicLines.length;
  const textStartY = 360;
  const available = 840 - textStartY;
  const totalH = totalLines * lineH;
  let y = textStartY + Math.max(0, (available - totalH) / 2);

  // Main lines – alternating white / gold
  ctx.font = mainFont;
  for (let i = 0; i < mainLines.length; i++) {
    const line = mainLines[i];
    const color = i % 2 === 1 ? rgb(GOLD) : rgb(WHITE);
    const lw = ctx.measureText(line).width;
    const x = (SIZE - lw) / 2;
    // Shadow
    ctx.fillStyle = rgba(0, 0, 0, 1);
    ctx.fillText(line, x + 3, y + 3);
    // Text
    ctx.fillStyle = color;
    ctx.fillText(line, x, y);
    y += lineH;
  }

  // Italic lines in gold
  ctx.font = italicFont;
  for (const line of italicLines) {
    const lw = ctx.measureText(line).width;
    const x = (SIZE - lw) / 2;
    ctx.fillStyle = rgba(0, 0, 0, 1);
    ctx.fillText(line, x + 3, y + 3);
    ctx.fillStyle = rgb(GOLD_LIGHT);
    ctx.fillText(line, x, y);
    y += lineH;
  }

  // ── CLOSING QUOTE MARK ──────────────────────────────────────────────────────
  ctx.font = qmFont;
  const cqmW = ctx.measureText("\u201d").width;
  ctx.fillStyle = rgb(GOLD);
  ctx.fillText("\u201d", (SIZE - cqmW) / 2, y + 105);

  // ── AUTHOR ──────────────────────────────────────────────────────────────────
  const authorY = y + 75;
  const cx2 = SIZE / 2;
  ctx.save();
  ctx.strokeStyle = rgba(GOLD[0], GOLD[1], GOLD[2], 0.63);
  ctx.lineWidth = 1;
  ctx.beginPath();
  ctx.moveTo(cx2 - 130, authorY);
  ctx.lineTo(cx2 + 130, authorY);
  ctx.stroke();
  ctx.restore();

  const authorStr = `\u2013 ${author.toUpperCase()}`;
  // Dynamische Schriftgrösse: startet bei 34px, reduziert bis 18px damit der Name nicht überläuft
  const MAX_AUTHOR_WIDTH = SIZE - 120; // 960px bei 1080px Canvas
  let authorFontSize = 34;
  ctx.font = fontSpec("Cinzel", authorFontSize);
  while (ctx.measureText(authorStr).width > MAX_AUTHOR_WIDTH && authorFontSize > 18) {
    authorFontSize -= 1;
    ctx.font = fontSpec("Cinzel", authorFontSize);
  }
  const aw = ctx.measureText(authorStr).width;
  const ax = (SIZE - aw) / 2;
  const ay = authorY + 18;
  // Shadow
  ctx.fillStyle = rgba(0, 0, 0, 1);
  ctx.fillText(authorStr, ax + 2, ay + 2);
  // Text
  ctx.fillStyle = rgb(WHITE);
  ctx.fillText(authorStr, ax, ay);

  // ── SAVE ────────────────────────────────────────────────────────────────────
  const pngBuf = canvas.toBuffer("image/png");
  await sharp(pngBuf).toFile(outputPath);
  console.log(`✅ Quote image saved: ${outputPath}`);
}

// CLI usage
if (process.argv[1] && process.argv[1].endsWith("createQuoteImage.ts")) {
  const [, , quote = "Test quote", author = "Test Author", output = "/tmp/quote_test_ts.png", bgUrl] = process.argv;
  createQuoteImage(quote, author, output, bgUrl).catch(console.error);
}
